uarray/backend.py

```diff
@@ -9,11 +9,62 @@
 ArgumentReplacerType = Callable[[Tuple, Dict, Tuple], Tuple[Tuple, Dict]]
 
 
-class BackendNotImplementedError(TypeError):
-    pass
+class BackendNotImplementedError(NotImplementedError):
+    """
+    An exception that is thrown when no compatible backend is found for a method.
+    """
 
 
 class MultiMethod:
+    """
+    This class is meant to provide a dispatchable method with or without a registered implementation.
+
+    Provides a multi-method instance. The method has no implementation by default, but one can be provided
+    if needed. :obj:`Backend` s can provide implementations for the method, and users can choose backends.
+
+    Parameters
+    ----------
+    argument_extractor : ArgumentExtractorType
+        This is a callable that extracts the arguments using which dispatch can be performed from this
+        multimethod.
+    argument_replacer : ArgumentReplacerType
+        This takes in args, kwargs and dispatchable args, and replaces all dispatchable arguments within
+        the args and kwargs, and then returns them.
+    default : Optional[Callable]
+        This is the default implementation for this multimethod, possibly in terms of other multimethods.
+
+    Examples
+    --------
+    >>> import uarray as ua
+    >>> def potato(a, b):
+    ...     return (a,) # b is not is dispatchable, so we return a only
+    >>> def potato_rd(args, kwargs, dispatch_args):
+    ...     # This replaces a within the args/kwargs
+    ...     return dispatch_args + args[1:], kwargs
+    >>> def potato_impl(a, b):
+    ...     # The default implementation passes through all arguments
+    ...     return a, b
+    >>> potato_mm = ua.MultiMethod(potato, potato_rd, default=potato_impl)
+    >>> potato_mm(1, '2')
+    (1, '2')
+    >>> be = ua.TypeCheckBackend((int,)) # Register implementation, define "supported dispatchable types"
```

Comment:
A little more text would help me here, something like: Register a second implementation, for which the supported type of `a` (I think, this is unclear - the rest of the example makes it look like it type checks `b`?) is only Python `int`.